src/api.rs

```diff
@@ -0,0 +1,243 @@
+        hub.save().await?;
+        Ok(old_name)
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn change_nickname(user: &User, hub_id: &ID, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    user.in_hub(&hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member_mut(&user.id)?;
+    let old_name = member.nickname.clone();
+    member.nickname = new_name;
+    hub.save().await?;
+    Ok(old_name)
+}
+
+pub async fn user_banned(user: &User, hub_id: &ID, user_id: &ID) -> Result<bool> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    Ok(hub.bans.contains(user_id))
+}
+
+pub async fn user_muted(user: &User, hub_id: &ID, user_id: &ID) -> Result<bool> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    Ok(hub.mutes.contains(user_id))
+}
+
+pub async fn get_hub_member(user: &User, hub_id: &ID, user_id: &ID) -> Result<HubMember> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if &user.id == user_id {
+        return Ok(member);
+    } else {
+        hub.get_member(user_id)
+    }
+}
+
+pub async fn join_hub(user: &mut User, hub_id: &ID) -> Result<()> {
+    user.join_hub(hub_id).await?;
+    user.save().await
+}
+
+pub async fn leave_hub(user: &mut User, hub_id: &ID) -> Result<()> {
+    user.leave_hub(hub_id).await?;
+    user.save().await
+}
+
+async fn hub_user_op(user: &User, hub_id: &ID, user_id: &ID, op: HubPermission) -> Result<()> {
+    user.in_hub(hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_permission(op.clone(), &hub) {
+        match op {
+            HubPermission::Kick => hub.kick_user(user_id).await?,
+            HubPermission::Ban => hub.ban_user(user_id.clone()).await?,
+            HubPermission::Unban => hub.unban_user(user_id),
+            HubPermission::Mute => hub.mute_user(user_id.clone()),
+            HubPermission::Unmute => hub.unmute_user(user_id),
+            _ => return Err(Error::UnexpectedServerArg),
+        }
+        hub.save().await
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn kick_user(user: &User, hub_id: &ID, user_id: &ID) -> Result<()> {
+    hub_user_op(user, hub_id, user_id, HubPermission::Kick).await
+}
+
+pub async fn ban_user(user: &User, hub_id: &ID, user_id: &ID) -> Result<()> {
+    hub_user_op(user, hub_id, user_id, HubPermission::Ban).await
+}
+
+pub async fn unban_user(user: &User, hub_id: &ID, user_id: &ID) -> Result<()> {
+    hub_user_op(user, hub_id, user_id, HubPermission::Unban).await
+}
+
+pub async fn mute_user(user: &User, hub_id: &ID, user_id: &ID) -> Result<()> {
+    hub_user_op(user, hub_id, user_id, HubPermission::Mute).await
+}
+
+pub async fn unmute_user(user: &User, hub_id: &ID, user_id: &ID) -> Result<()> {
+    hub_user_op(user, hub_id, user_id, HubPermission::Unmute).await
+}
```

Comment:
might be better to just use a macro to generate these functions with the same signature.
```suggestion
macro_rules! action_fns {
  ($(($fnName:ident, $variant:ident)),*) => {
    $(
      pub async fn $fnName(user: &User, hub_id: &ID, user_id: &ID) -> Result<()> {
          hub_user_op(user, hub_id, user_id, HubPermission::$variant).await
      }
    )*
  }
}
action_fns! { 
(kick_user, Kick), 
(ban_user, Ban), 
(unban_user, Unban), 
(mute_user, Mute),
(unmute_user, Unmute)
}
```
you can still keep these if you like.

Reply:
I thought those were messy and wanted to clean them up, I only kept them separate because of this being a public API, good to know that I can do this to clean it up in the actual code.